src/core/credentials/transport/xds/xds_credentials.cc

```diff
@@ -73,28 +92,42 @@ bool XdsVerifySubjectAlternativeNames(
 //
 
 XdsCertificateVerifier::XdsCertificateVerifier(
-    RefCountedPtr<XdsCertificateProvider> xds_certificate_provider)
-    : xds_certificate_provider_(std::move(xds_certificate_provider)) {}
+    RefCountedPtr<XdsCertificateProvider> xds_certificate_provider,
+    std::string sni_name)
+    : xds_certificate_provider_(std::move(xds_certificate_provider)),
+      sni_matcher_(
+          StringMatcher::Create(StringMatcher::Type::kExact, sni_name, true)
+              .value()) {}
 
 bool XdsCertificateVerifier::Verify(
     grpc_tls_custom_verification_check_request* request,
     std::function<void(absl::Status)>, absl::Status* sync_status) {
   GRPC_CHECK_NE(request, nullptr);
-  if (!XdsVerifySubjectAlternativeNames(
-          request->peer_info.san_names.uri_names,
-          request->peer_info.san_names.uri_names_size,
-          xds_certificate_provider_->san_matchers()) &&
-      !XdsVerifySubjectAlternativeNames(
-          request->peer_info.san_names.ip_names,
-          request->peer_info.san_names.ip_names_size,
-          xds_certificate_provider_->san_matchers()) &&
-      !XdsVerifySubjectAlternativeNames(
-          request->peer_info.san_names.dns_names,
-          request->peer_info.san_names.dns_names_size,
-          xds_certificate_provider_->san_matchers())) {
-    *sync_status = absl::Status(
-        absl::StatusCode::kUnauthenticated,
-        "SANs from certificate did not match SANs from xDS control plane");
+  if (xds_certificate_provider_->auto_sni_san_validation()) {
+    if (!XdsVerifySubjectAlternativeNames(
+            request->peer_info.san_names.dns_names,
+            request->peer_info.san_names.dns_names_size, {sni_matcher_})) {
+      *sync_status = absl::Status(
```

Comment:
Instead of `absl::Status(absl::StatusCode::kUnauthenticated, "foo")`, please say `absl::UnauthenticatedError("foo")`.  (I'm not sure why the original code was written this way.)

Same thing on line 127.